File: python/quicksort.py
import unittest, random

def quicksort(l):
    if len(l) <= 1:
        return l
    sort(l, 0, len(l))
    return l

def sort(l, low, high):
    if high - low > 1:
        # pick random pivot
        pivot_idx = random.randrange(low, high)
        # swap pivot to rightmost spot
        pivot = l[pivot_idx]
        l[pivot_idx] = l[high - 1]
        l[high - 1] = pivot

        partition = low
        for i in range(low, high - 1):
            if l[i] < pivot:
                temp = l[i]
                l[i] = l[partition]
                l[partition] = temp
                partition += 1
        l[high - 1] = l[partition]
        l[partition] = pivot

        sort(l, low, partition)
        sort(l, partition + 1, high)

class TestQuickSort(unittest.TestCase):

    def test_empty_array(self):
        l = []
        result = quicksort(l)
        self.assertEqual(result, [])

    def test_single_elem(self):
        l = [6]
        result = quicksort(l)
        self.assertEqual(result, [6])

    def test_two_elems(self):
        l = [3, 0]
        result = quicksort(l)
        self.assertEqual(result, [0, 3])

    def test_three_elems(self):
        l = [3, 2, 1]
        result = quicksort(l)
        self.assertEqual(result, [1, 2, 3])

    def test_many_elems(self):
        l = [9, 1, 2, 6, 7, 8, 10, 11, 23, 90, 5, 9]
        result = quicksort(l)
        self.assertEqual(result, [1, 2, 5, 6, 7, 8, 9, 9, 10, 11, 23, 90])


if __name__ == "__main__":
    unittest.main()

File: rust/src/quicksort.rs
use rand::prelude::*;

fn quicksort(mut arr: &mut Vec<u32>, length: usize) {
    quicksort_helper(arr, 0, length);
}

fn quicksort_helper(mut arr: &mut Vec<u32>, low: usize, high: usize) {
    if high - low > 1 {
        // pick a pivot and swap with rightmost element
        let mut rng = thread_rng();
        let p_idx: usize = rng.gen_range(low, high);
        let pivot = arr[p_idx];
        arr[p_idx] = arr[high - 1];
        arr[high - 1] = pivot;
        let mut partition: usize = low;

        for i in low..high - 1 {
            if arr[i] < pivot {
                let temp = arr[i];
                arr[i] = arr[partition];
                arr[partition] = temp;
                partition += 1;
            }
        }
        arr[high - 1] = arr[partition];
        arr[partition] = pivot;

        quicksort_helper(arr, low, partition);
        quicksort_helper(arr, partition + 1, high);
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn empty_vec() {
        let mut arr = Vec::new();
        let len = arr.len();
        quicksort(&mut arr, len);
        assert_eq!(arr, vec![]);
    }

    #[test]
    fn one_item() {
        let mut arr = vec![1];
        let len = arr.len();
        quicksort(&mut arr, len);
        assert_eq!(arr, vec![1]);
    }

    #[test]
    fn reverse() {
        let mut arr = vec![5, 4, 2, 1, 0];
        let len = arr.len();
        quicksort(&mut arr, len);
        assert_eq!(arr, vec![0, 1, 2, 4, 5]);
    }

    #[test]
    fn scrambled() {
        let mut arr = vec![4, 2, 5, 1, 0];
        let len = arr.len();
        quicksort(&mut arr, len);
        assert_eq!(arr, vec![0, 1, 2, 4, 5]);
    }

}
